File: bridge_login.py
# -*- coding: utf-8 -*-
"""
Login a registered Postman account into the bridge's Chrome profile (.postman-profile).
After this, the TS bridge (Playwright) can reuse the logged-in session.

Usage:
  python bridge_login.py [email] [password]
  (defaults: newest account from postman_accounts.jsonl)
"""
import asyncio
import json
import os
import sys
import time

sys.stdout.reconfigure(encoding="utf-8", errors="replace")
import nodriver as uc

PROFILE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), ".postman-profile")
ACCOUNTS_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "postman_accounts.jsonl")
CHECK_DIR = r"D:\hoat_hinh\Grokpool\check"


def newest_account():
    if not os.path.exists(ACCOUNTS_FILE):
        return None
    with open(ACCOUNTS_FILE, "r", encoding="utf-8") as f:
        lines = [l.strip() for l in f if l.strip()]
    if not lines:
        return None
    # last line = newest
    return json.loads(lines[-1])


async def js(tab, script):
    try:
        r = await tab.evaluate(script)
        return "" if r is None else str(r)
    except Exception:
        return ""


async def wait_cf(tab, max_wait=60):
    for i in range(max_wait // 3):
        await asyncio.sleep(3)
        body = await js(tab, "document.body.innerText") or ""
        url = await js(tab, "window.location.href") or ""
        if not ("Performing security verification" in body or "Verify you are human" in body or "Just a moment" in body or "challenges.cloudflare" in url):
            return True
    return False


async def wait_turnstile(tab, max_wait=45):
    for i in range(max_wait // 2):
        await asyncio.sleep(2)
        v = await js(tab, "(() => { const r = document.querySelector(\"input[name='cf-turnstile-response']\"); return (r && r.value && r.value.length > 10) ? 'ready' : 'no'; })()")
        if v == "ready":
            return True
        body = await js(tab, "document.body.innerText") or ""
        if "Unable to verify" in body:
            return False
    return False


async def main():
    account = newest_account()
    if len(sys.argv) >= 3:
        account = {"email": sys.argv[1], "password": sys.argv[2]}
    if not account:
        print("no account found")
        return 1

    email = account["email"]
    password = account["password"]
    print(f"login: {email}")

    browser = await uc.start(
        headless=False,
        browser_args=[
            "--disable-gpu",
            f"--user-data-dir={PROFILE_DIR}",
            "--window-size=1280,900",
        ],
    )

    tab = await browser.get("https://identity.getpostman.com/login")

    # CF
    if not await wait_cf(tab):
        print("CF FAILED")
        return 1

    # DOM
    for i in range(20):
        await asyncio.sleep(2)
        n = await js(tab, "document.querySelectorAll('input:not([type=hidden]):not([type=checkbox])').length")
        if n.isdigit() and int(n) >= 2:
            break

    # Turnstile
    if not await wait_turnstile(tab):
        print("Turnstile FAILED")
        return 1

    # Fill
    await js(tab, f"""
        (() => {{
            const u = document.querySelector('#username');
            const p = document.querySelector('#password');
            if (u) {{ u.focus(); u.value = '{email}'; u.dispatchEvent(new Event('input', {{bubbles:true}})); }}
            if (p) {{ p.focus(); p.value = '{password}'; p.dispatchEvent(new Event('input', {{bubbles:true}})); }}
            return 'ok';
        }})()
    """)
    await asyncio.sleep(1)
    await js(tab, "document.querySelector('button[type=submit]').click()")
    print("submitted, waiting for redirect...")

    # Wait to land on postman workspace
    for i in range(40):
        await asyncio.sleep(3)
        url = await js(tab, "window.location.href") or ""
        body = await js(tab, "document.body.innerText") or ""
        if "postman.co" in url and "identity.getpostman.com" not in url:
            print(f"LOGGED IN: {url[:100]}")
            # Wait a bit for the app to save session cookies
            await asyncio.sleep(10)
            try:
                await tab.save_screenshot(os.path.join(CHECK_DIR, "bridge_login_ok.png"))
            except Exception:
                pass
            browser.stop()
            # save login url for .env
            with open(os.path.join(os.path.dirname(ACCOUNTS_FILE), "bridge_workspace_url.txt"), "w") as f:
                f.write(url)
            return 0
        if "incorrect" in body.lower() or "invalid" in body.lower():
            print(f"LOGIN ERROR: {body[:200]}")
            browser.stop()
            return 1
    print("login redirect timeout")
    browser.stop()
    return 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: scripts/bridge-login.mjs
/**
 * Login Postman into the bridge's Playwright profile (.postman-profile).
 * Uses the same engine (playwright-core + same profile) as the bridge itself.
 * Usage: node scripts/bridge-login.mjs <email> <password>
 */
import path from 'node:path';
import { chromium } from 'playwright-core';

const [email, password] = process.argv.slice(2);
if (!email || !password) {
  console.error('usage: node scripts/bridge-login.mjs <email> <password>');
  process.exit(1);
}

const profileDir = path.resolve('.postman-profile');
console.log(`profile: ${profileDir}`);

function findChrome() {
  const candidates = [
    'C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe',
    'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe',
  ];
  for (const c of candidates) {
    try { if (require('node:fs').existsSync(c)) return c; } catch {}
  }
  return undefined;
}

const context = await chromium.launchPersistentContext(profileDir, {
  headless: false,
  channel: 'chrome',
  viewport: { width: 1280, height: 900 },
  args: ['--disable-blink-features=AutomationControlled', '--disable-gpu'],
});

const page = context.pages()[0] ?? (await context.newPage());
page.setDefaultTimeout(45_000);

// 1. Open login page
console.log('navigating to login...');
await page.goto('https://identity.getpostman.com/login', { waitUntil: 'domcontentloaded', timeout: 60_000 }).catch(() => {});

// 2. Wait for CF to clear
for (let i = 0; i < 30; i++) {
  const body = (await page.locator('body').innerText().catch(() => '')) || '';
  if (!/performing security verification|verify you are human|just a moment/i.test(body)) {
    console.log('CF clear');
    break;
  }
  await page.waitForTimeout(3000);
}

// If already logged in, this will redirect to workspace
await page.waitForTimeout(3000);
let url = page.url();
console.log(`url now: ${url.slice(0, 90)}`);

if (/identity\.getpostman\.com\/login/.test(url)) {
  // 3. Fill the form
  console.log('filling login form...');
  await page.waitForSelector('#username', { timeout: 30_000 });
  await page.fill('#username', email);
  await page.fill('#password', password);

  // 4. Wait for turnstile to be ready
  for (let i = 0; i < 25; i++) {
    const ready = await page.evaluate(() => {
      const el = document.querySelector("input[name='cf-turnstile-response']");
      return !!(el && el.value && el.value.length > 10);
    });
    if (ready) { console.log('turnstile ready'); break; }
    await page.waitForTimeout(2000);
  }

  // 5. Submit
  await page.click('button[type=submit]');
  console.log('submitted');
}

// 6. Wait for postman.co redirect
let loggedIn = false;
for (let i = 0; i < 40; i++) {
  await page.waitForTimeout(3000);
  url = page.url();
  if (url.includes('postman.co') && !url.includes('identity.getpostman.com')) {
    loggedIn = true;
    console.log(`LOGGED IN: ${url.slice(0, 100)}`);
    break;
  }
  const body = (await page.locator('body').innerText().catch(() => '')) || '';
  if (/incorrect|invalid credentials/i.test(body.slice(0, 3000))) {
    console.error('LOGIN ERROR: wrong credentials');
    await context.close();
    process.exit(1);
  }
}

if (!loggedIn) {
  console.error('login redirect timeout');
  await page.screenshot({ path: '.runtime/login-timeout.png' }).catch(() => {});
  await context.close();
  process.exit(1);
}

// 7. Wait for the app to fully initialize + persist cookies
await page.waitForTimeout(10_000);

// Keep browser open briefly to let Chromium flush profile to disk
await context.close();
console.log('done - session saved to profile');
